AI freshness April: update evaluation and tokenizer docs (#52961)
dotnet/docs: docs/ai/how-to/use-tokenizers.md

title: Use Microsoft.ML.Tokenizers for text tokenization
description: Learn how to use the Microsoft.ML.Tokenizers library to tokenize text for AI models, manage token counts, and work with various tokenization algorithms.
ms.topic: how-to
ms.date: 04/09/2026
ai-usage: ai-assisted
---
# Use Microsoft.ML.Tokenizers for text tokenization

> The [Microsoft.ML.Tokenizers](https://www.nuget.org/packages/Microsoft.ML.Tokenizers) library provides a comprehensive set of tools for tokenizing text in .NET applications. Tokenization is essential when working with large language models (LLMs), as it lets you manage token counts, estimate costs, and preprocess text for AI models.

Reviewer: ```suggestion
The [Microsoft.ML.Tokenizers](https://www.nuget.org/packages/Microsoft.ML.Tokenizers) library provides a comprehensive set of tools for tokenizing text in .NET applications. Tokenization is essential when you work with large language models (LLMs), as it lets you manage token counts, estimate costs, and preprocess text for AI models.
```

:::code language="csharp" source="./snippets/use-tokenizers/csharp/TokenizersExamples/TiktokenExample.cs" id="TiktokenBasic":::

> Cache and reuse the tokenizer instance for better performance.

Reviewer: ```suggestion
For better performance, cache and reuse the tokenizer instance throughout your app.
```

> Working with LLMs often requires managing text within token limits. The following example shows how to trim text to a specific token count:

Reviewer: ```suggestion
When you work with LLMs, you often need to manage text within token limits. The following example shows how to trim text to a specific token count:
```

## Use BPE tokenizer

> *Byte-pair encoding* (BPE) is the underlying algorithm used by many tokenizers, including Tiktoken. BPE was initially developed as an algorithm to compress texts, and then OpenAI used it for tokenization when pretraining the GPT model. The following example demonstrates BPE tokenization:

Reviewer: ```suggestion
*Byte-pair encoding* (BPE) is the underlying algorithm used by many tokenizers, including Tiktoken. BPE was initially developed as an algorithm to compress texts, and then OpenAI used it for tokenization when it pretrained the GPT model. The following example demonstrates BPE tokenization:
```
